### PR TITLE
Feature: limit to root pages

### PR DESCRIPTION
**Pull request by mvscheidt**
- added three new dca fields (`limitToRootPages`, `rootPageIds`, `negateRootPageIds`)
- added ModuleFieldsListener class for `options_callback` of `rootPageIds`
- added constructor to `PageFinder` with two optional parameters (to avoid breaking already existing code) - `$rootPageIds` for an array of root ids and `$negateRootPageIds` which will negate the sql condition (`IN()` to `NOT IN()`)
- added method `addLimitRootPageIdsCondition` to `PageFinder` which adds the `IN()` or `NOT IN()` condition to the methods `findRootPagesForPage` and `findMasterRootForPage`
- added config handling to `ChangeLanguageModule` to instance `PageFinder` with the new config settings - the class also has a `getRootPageIds` method because based on the config we may need to add or remove the rootId of the current page (see docblocks for details)
- added unittests for
  - limiting rootPages (only the choosen root pages are shown in the FE language navigation)
  - negating the list of rootPages (only pages NOT in the list are shown in the FE language navigation)

## ToDo

- [x] Talk to terminal42 if they are interested in this as PR (https://github.com/terminal42/contao-changelanguage/issues/135)
  - ~~If yes Talk to terminal42 about Transifex translations and how they want them done (currently missing)~~
  - [x] If no, we will just add the translations manually to xlf files in our repo

### CODE REVIEW

**Comment by mvscheidt** [comment]
If you have better naming for any of the new dca fields say so. you know I suck at naming things

File: library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php
Comment (mvscheidt) [inline library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php]: I've been thinking some more, maybe we should move both of these conditions into the `PageFinder` instead. But I am torn on this.

On one hand the way it's currently done modules can decide themselves how you want to handle the "current page" root (if you want it included / excluded based on the config).

On the other hand it maybe makes more sense to just pass rootIds (example: `1,25,54`) and then the PageFinder can add / remove a rootId in its list based on the config  (depends on the "current" Page that has been passed to `findRootPagesForPage` or `findMasterRootForPage`)

**Comment by mvscheidt** [comment]
I think we can do this in a small contao extension / contao-bundle. There is a Hook `changelanguageNavigation` I missed: https://github.com/terminal42/contao-changelanguage/issues/135#issuecomment-326018632

Only thing I have to think about is how / where we place the config (needs to be configurable in the backend).

Another Reason why we should use the hook - since it won't be merged into the original repository (our use case is to rare - never came up until now) we would have to update our fork if the original repository updates and also maintain our changes ontop of that.

File: dca/tl_module.php
Comment (DanielSchwiperich) [inline dca/tl_module.php]: I think `limitWebsiteRoots` would be a better naming (since Website root is used as name by contao itself)

File: dca/tl_module.php
Comment (DanielSchwiperich) [inline dca/tl_module.php]: also then change to `negateWebsiteRootsSelection`

File: library/Terminal42/ChangeLanguage/PageFinder.php
Comment (DanielSchwiperich) [inline library/Terminal42/ChangeLanguage/PageFinder.php]: "$table.id ${negation} IN(${placeholders})"

should be this way imho

File: library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php
Comment (DanielSchwiperich) [reply library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php]: I think you right. The page finder should be responsible for this

**Review by DanielSchwiperich — CHANGES_REQUESTED** [review]
Some renaming needs to be done and some small changes

File: library/Terminal42/ChangeLanguage/PageFinder.php
Comment (mvscheidt) [reply library/Terminal42/ChangeLanguage/PageFinder.php]: This looked so weird to me I googled it. I found my answer and found this interesting tidbit aswell regarding performance of strings/concatenation : http://php.net/manual/de/language.types.string.php#120160

**Comment by mvscheidt** [comment]
By the way - should we really merge this into master @DanielSchwiperich - I'm thinking if we ever find some other issue we cannot use our master anymore to create pull requests for the original repo.

Not sure what's best practise on github for that

File: library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php
Comment (mvscheidt) [reply library/Terminal42/ChangeLanguage/FrontendModule/ChangeLanguageModule.php]: see ad3a9ef

File: dca/tl_module.php
Comment (mvscheidt) [reply dca/tl_module.php]: see b31236f

File: dca/tl_module.php
Comment (mvscheidt) [reply dca/tl_module.php]: see b31236f

File: library/Terminal42/ChangeLanguage/PageFinder.php
Comment (mvscheidt) [reply library/Terminal42/ChangeLanguage/PageFinder.php]: see c3f098b

**Comment by DanielSchwiperich** [comment]
we could keep the branch and create a tag / version from it. for future updates we then update master and rebase the branch

**Comment by mvscheidt** [comment]
I rebased on current master so we also get our fix for the sorting issue when we use our featurebranch in our project.

Also remember to NOT merge this PR after review - we will just close it.

**Review by DanielSchwiperich — APPROVED** [review]
There's mixed use of string concatenation, but it's ok for me

**Comment by mvscheidt** [comment]
Ok I will close the merge request then so nobody will merge it by accident.

**Pull request closed without merging**
